src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +558,109 @@ ApplicationImpl::joinAllThreads()
     LOG(DEBUG) << "Joined all " << mWorkerThreads.size() << " threads";
 }
 
-bool
-ApplicationImpl::manualClose()
+std::string
+ApplicationImpl::manualClose(optional<uint32_t> const& manualLedgerSeq,
+                             optional<TimePoint> const& manualCloseTime)
 {
     // Manual close only makes sense for validating nodes
     if (!mConfig.NODE_IS_VALIDATOR)
     {
-        return false;
+        throw std::invalid_argument(
+            "Issuing a manual ledger close requires NODE_IS_VALIDATOR to be "
+            "set to true.");
     }
 
     if (mConfig.MANUAL_CLOSE)
     {
-        mHerder->triggerNextLedger(mLedgerManager->getLastClosedLedgerNum() + 1,
-                                   true);
-        return true;
+        auto const startLedgerSeq = getLedgerManager().getLastClosedLedgerNum();
+        if (startLedgerSeq == std::numeric_limits<uint32_t>::max())
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Manually closed ledger "
+                                       "sequence number already at max ({})"),
+                            startLedgerSeq));
+        }
+        auto minLedgerSeq = startLedgerSeq + 1;
+        auto const targetLedgerSeq =
+            manualLedgerSeq ? *manualLedgerSeq : minLedgerSeq;
+        if (targetLedgerSeq < minLedgerSeq)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Invalid manual close ledger "
+                                       "sequence number {} (minimum {})"),
+                            targetLedgerSeq, minLedgerSeq));
+        }
+        if (!mHerder->isTransactionQueueEmpty() &&
```

Comment:
> Why do we need this check? Because we already use the normal code when triggering ledgers we'll do the right thing with any potential transaction that gets invalidated by the new close time / ledger seq

When we manually close a ledger, there might be more transactions in the queue than would fit in a single ledger.  So to skip ledgers in that state, we have to close ledgers one at a time until the queue is empty, to get the right semantics regarding which transactions get into which ledgers (which may have further consequences).

So if we want to skip ahead, we have to choose where to do the one-by-one closing to empty the transaction queue first:  the client (usually Horizon), or the core.  The core could do it, or it could return an error and leave it to the client.  I think that it'd be basically the same code in either place, and I gather it's a generally-agreed principle (it would certainly make sense to me) that, all other things being equal, complexity is better outside `stellar-core` than inside it, so I went with leaving it to the client.

I could stick the loop in `stellar-core` itself if you prefer.

Reply:
not really: this is an arbitrary rule that should be removed.

In a real deployment of stellar-core, the decisions made wrt the transaction queue by the local node are pretty much independent of whatever happens during consensus. In fact this kind of thing happens pretty often as nodes on the network are not closing ledgers at exactly the same time but still accept/flood transactions.

Like you said, we should leave the decision on what to simulate to the caller, and this rule gets in the way.

Reply:
> In a real deployment of stellar-core, the decisions made wrt the transaction queue by the local node are pretty much independent of whatever happens during consensus. In fact this kind of thing happens pretty often as nodes on the network are not closing ledgers at exactly the same time but still accept/flood transactions.

This is what I had failed to understand:  we might close a bunch of ledgers without taking any transactions from our queue, as it's consensus that dictates which transactions get in.  I had been stuck in "standalone mode" myself, thinking that the node would deterministically take some given set of transactions from its queue on each close.

I've removed `isTransactionQueueEmpty()`.